From c56630916167dbce3243cf68f8b08914c7875e3e Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 187/296] Avoid dangerous recursive call on startProcessLocked

Symptom:
StackOverflow happened at ActivityManagerService#startProcessLocked().

When user data for the current home is cleared, the home package
goes to frozen state in a short time. Home process stops and restarts.
At that time, StackOverflow happens sometimes.

Root cause:
startProcessLocked() can be failed if the target package is frozen.
It has a recovery logic for failure. It calls forceStopPackageLocked()
and it makes unwanted recursive call.

Solution:
Calling forceStopPackageLocked from outside of call stack.
---
 .../core/java/com/android/server/am/ActivityManagerService.java   | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/services/core/java/com/android/server/am/ActivityManagerService.java b/services/core/java/com/android/server/am/ActivityManagerService.java
index cb6ad206a14..031297e04f3 100644
--- a/services/core/java/com/android/server/am/ActivityManagerService.java
+++ b/services/core/java/com/android/server/am/ActivityManagerService.java
@@ -4089,8 +4089,12 @@ public class ActivityManagerService extends IActivityManager.Stub
             // starting this process. (We already invoked this method once when
             // the package was initially frozen through KILL_APPLICATION_MSG, so
             // it doesn't hurt to use it again.)
-            forceStopPackageLocked(app.info.packageName, UserHandle.getAppId(app.uid), false,
-                    false, true, false, false, UserHandle.getUserId(app.userId), "start failure");
+            mHandler.post(() -> {
+                synchronized(ActivityManagerService.this) {
+                    forceStopPackageLocked(app.info.packageName, UserHandle.getAppId(app.uid), false,
+                        false, true, false, false, UserHandle.getUserId(app.userId), "start failure");
+                }
+            });
         }
     }
 
-- 
2.11.0

